### ISSUE TITLE
Mask R-CNN with deeplearning4j

### ISSUE THREAD

**Issue by fpeapptus** [eclipse/deeplearning4j#6616]
Hi. We tried to make a simple Mask R-CNN project (for finding particular clothing items in pictures) and use the result in Java with deeplearning4j. We have made a keras model with python that works and now we are trying to import it with the following code: 
```
        String simpleMlp = new ClassPathResource("saved_model.hdf5").getFile().getPath();
        ComputationGraph model = KerasModelImport.importKerasModelAndWeights(simpleMlp);
```
org.deeplearning4j.nn.modelimport.keras.exceptions.UnsupportedKerasConfigurationException: Unsupported keras layer type BatchNorm. Am I doing something wrong or is this simply not supported (yet?).

**Comment by AlexDBlack** [comment]
Ok, I'm not 100% sure, but you *might* be running into this issue: https://github.com/deeplearning4j/deeplearning4j/issues/6550

Specifically there was a JSON format change in Keras 2.2.4 that has been fixed on snapshots: https://deeplearning4j.org/docs/latest/deeplearning4j-config-snapshots

So you can try snapshots - beyond that, can you share your model configuration file for us to test locally to reproduce (or confirm as fixed)? Thanks.

**Comment by fpeapptus** [comment]
I tried running it again with versions 1.0.0-SNAPSHOT and I got the same exception. I don't have the model configuration file, just the model and it is too big to upload. You can try with an example R-CNN project to reproduce maybe, we didn't make any notable changes. We decided to go with just a tensor flow dependency and have made some progress there.

**Comment by girishnr1** [comment]
I'm having similar issue loading .h5 trained uisng VGG16,  using keras 2.2.4.
INFO org.deeplearning4j.nn.modelimport.keras.Hdf5Archive - Unexpected end-of-input in FIELD_NAME
Tried using 1.0.0-SNAPSHOT, 1.0.0-beta2, 1.0.0-beta3, same exception in all 3.
Any updated on this issue?  Thanks

**Comment by girishnr1** [comment]
Looks like above messages are benign [ Unexpected end-of-input in FIELD_NAME].  

The exception is thrown during buildMode() org.deeplearning4j.nn.modelimport.keras.exceptions.UnsupportedKerasConfigurationException: Unsupported keras layer type Model.  Thanks

**Comment by saudet** [comment]
Looks like a bug in GCC:
> // Some platforms will occasionally send nonstandard spurious SI_KERNEL codes.
> // One way to get this is via unaligned SIMD loads.
- https://reviews.llvm.org/D12659

Unaligned memory access easily happens with views, but I suppose it's the compiler's job to make sure that doesn't happen when we're not calling any intrinsics manually.

**Comment by maxpumperla** [comment]
@fpeapptus if you have the model, why can't you quickly load it back to keras, pump out the json configuration and share it with us? Also, can you please share your full stack trace with the error message you're seeing currently? That would greatly help. As @AlexDBlack mentioned, batch norm has been supported since 0.9.1, so I don't believe it until I see it.

@girishnr1 your error is completely unrelated, but a duplicate of https://github.com/deeplearning4j/deeplearning4j/issues/5411.

**Comment by girishnr1** [comment]
Thanks for the response.  Please find stack trace below:

[org.deeplearning4j.nn.modelimport.keras.utils.KerasLayerUtils.getKerasLayerFromConfig(KerasLayerUtils.java:318), org.deeplearning4j.nn.modelimport.keras.KerasModel.prepareLayers(KerasModel.java:218), org.deeplearning4j.nn.modelimport.keras.KerasModel.<init>(KerasModel.java:164), org.deeplearning4j.nn.modelimport.keras.KerasModel.<init>(KerasModel.java:96), org.deeplearning4j.nn.modelimport.keras.utils.KerasModelBuilder.buildModel(KerasModelBuilder.java:307), com.klatencor.cougar.App.loadModel(App.java:50), com.klatencor.cougar.App.main(App.java:26)


Also, please find JSON saved through 'model.to_json()' API.

Thanks

{"class_name": "Model", "config": {"name": "model_1", "layers": [{"name": "image_input", "class_name": "InputLayer", "config": {"batch_input_shape": [null, 224, 224, 3], "dtype": "float32", "sparse": false, "name": "image_input"}, "inbound_nodes": []}, {"name": "vgg16", "class_name": "Model", "config": {"name": "vgg16", "layers": [{"name": "input_1", "class_name": "InputLayer", "config": {"batch_input_shape": [null, null, null, 3], "dtype": "float32", "sparse": false, "name": "input_1"}, "inbound_nodes": []}, {"name": "block1_conv1", "class_name": "Conv2D", "config": {"name": "block1_conv1", "trainable": false, "filters": 64, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["input_1", 0, 0, {}]]]}, {"name": "block1_conv2", "class_name": "Conv2D", "config": {"name": "block1_conv2", "trainable": false, "filters": 64, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block1_conv1", 0, 0, {}]]]}, {"name": "block1_pool", "class_name": "MaxPooling2D", "config": {"name": "block1_pool", "trainable": false, "pool_size": [2, 2], "padding": "valid", "strides": [2, 2], "data_format": "channels_last"}, "inbound_nodes": [[["block1_conv2", 0, 0, {}]]]}, {"name": "block2_conv1", "class_name": "Conv2D", "config": {"name": "block2_conv1", "trainable": false, "filters": 128, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block1_pool", 0, 0, {}]]]}, {"name": "block2_conv2", "class_name": "Conv2D", "config": {"name": "block2_conv2", "trainable": false, "filters": 128, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block2_conv1", 0, 0, {}]]]}, {"name": "block2_pool", "class_name": "MaxPooling2D", "config": {"name": "block2_pool", "trainable": false, "pool_size": [2, 2], "padding": "valid", "strides": [2, 2], "data_format": "channels_last"}, "inbound_nodes": [[["block2_conv2", 0, 0, {}]]]}, {"name": "block3_conv1", "class_name": "Conv2D", "config": {"name": "block3_conv1", "trainable": false, "filters": 256, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block2_pool", 0, 0, {}]]]}, {"name": "block3_conv2", "class_name": "Conv2D", "config": {"name": "block3_conv2", "trainable": false, "filters": 256, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block3_conv1", 0, 0, {}]]]}, {"name": "block3_conv3", "class_name": "Conv2D", "config": {"name": "block3_conv3", "trainable": false, "filters": 256, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block3_conv2", 0, 0, {}]]]}, {"name": "block3_pool", "class_name": "MaxPooling2D", "config": {"name": "block3_pool", "trainable": false, "pool_size": [2, 2], "padding": "valid", "strides": [2, 2], "data_format": "channels_last"}, "inbound_nodes": [[["block3_conv3", 0, 0, {}]]]}, {"name": "block4_conv1", "class_name": "Conv2D", "config": {"name": "block4_conv1", "trainable": false, "filters": 512, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block3_pool", 0, 0, {}]]]}, {"name": "block4_conv2", "class_name": "Conv2D", "config": {"name": "block4_conv2", "trainable": false, "filters": 512, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block4_conv1", 0, 0, {}]]]}, {"name": "block4_conv3", "class_name": "Conv2D", "config": {"name": "block4_conv3", "trainable": false, "filters": 512, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block4_conv2", 0, 0, {}]]]}, {"name": "block4_pool", "class_name": "MaxPooling2D", "config": {"name": "block4_pool", "trainable": false, "pool_size": [2, 2], "padding": "valid", "strides": [2, 2], "data_format": "channels_last"}, "inbound_nodes": [[["block4_conv3", 0, 0, {}]]]}, {"name": "block5_conv1", "class_name": "Conv2D", "config": {"name": "block5_conv1", "trainable": false, "filters": 512, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block4_pool", 0, 0, {}]]]}, {"name": "block5_conv2", "class_name": "Conv2D", "config": {"name": "block5_conv2", "trainable": false, "filters": 512, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block5_conv1", 0, 0, {}]]]}, {"name": "block5_conv3", "class_name": "Conv2D", "config": {"name": "block5_conv3", "trainable": false, "filters": 512, "kernel_size": [3, 3], "strides": [1, 1], "padding": "same", "data_format": "channels_last", "dilation_rate": [1, 1], "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["block5_conv2", 0, 0, {}]]]}, {"name": "block5_pool", "class_name": "MaxPooling2D", "config": {"name": "block5_pool", "trainable": false, "pool_size": [2, 2], "padding": "valid", "strides": [2, 2], "data_format": "channels_last"}, "inbound_nodes": [[["block5_conv3", 0, 0, {}]]]}], "input_layers": [["input_1", 0, 0]], "output_layers": [["block5_pool", 0, 0]]}, "inbound_nodes": [[["image_input", 0, 0, {}]]]}, {"name": "flatten", "class_name": "Flatten", "config": {"name": "flatten", "trainable": true, "data_format": "channels_last"}, "inbound_nodes": [[["vgg16", 1, 0, {}]]]}, {"name": "fc1", "class_name": "Dense", "config": {"name": "fc1", "trainable": true, "units": 512, "activation": "linear", "use_bias": true, "kernel_initializer": {"class_name": "RandomNormal", "config": {"mean": 0.0, "stddev": 0.05, "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["flatten", 0, 0, {}]]]}, {"name": "dropout_1", "class_name": "Dropout", "config": {"name": "dropout_1", "trainable": true, "rate": 0.5, "noise_shape": null, "seed": null}, "inbound_nodes": [[["fc1", 0, 0, {}]]]}, {"name": "fc2", "class_name": "Dense", "config": {"name": "fc2", "trainable": true, "units": 256, "activation": "linear", "use_bias": true, "kernel_initializer": {"class_name": "RandomNormal", "config": {"mean": 0.0, "stddev": 0.05, "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["dropout_1", 0, 0, {}]]]}, {"name": "dropout_2", "class_name": "Dropout", "config": {"name": "dropout_2", "trainable": true, "rate": 0.5, "noise_shape": null, "seed": null}, "inbound_nodes": [[["fc2", 0, 0, {}]]]}, {"name": "predictions", "class_name": "Dense", "config": {"name": "predictions", "trainable": true, "units": 2, "activation": "linear", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["dropout_2", 0, 0, {}]]]}], "input_layers": [["image_input", 0, 0]], "output_layers": [["predictions", 0, 0]]}, "keras_version": "2.2.4", "backend": "tensorflow"}

**Comment by maxpumperla** [comment]
@girishnr1 again, your issue is unrelated. you have a nested `Model`, which we don't support yet. please stop spamming in unrelated issues.

@fpeapptus if you could please provide your model json and stack trace, I'm happy to take a look.

**Comment by maxpumperla** [comment]
I'm leaving this issue open as feature request, because I think the application is really cool and we should think about supporting it at some point. Just keep in mind that in the above `model.py` is really just the first of many custom layers this model uses. So this is hard work and very low in priority for us.

**Comment by agibsonccc** [comment]
We support custom layer import now. Closing this.
